From fe748cf464935c6bb6e54288f4d9dcabdf9689ea Mon Sep 17 00:00:00 2001
From: Dominik Haller <d.haller@phytec.de>
Date: Mon, 18 Sep 2023 10:28:48 +0200
Subject: [HACK phyboard-izar-am68x-3 1/2] HACK: board: phytec: phycore_am68x:
 Obtain RAM size from devicetree

Utilize the fdtdec_setup* functions to obtain the RAM size from the
devicetree.

Signed-off-by: Dominik Haller <d.haller@phytec.de>
---
 .../dts/k3-am68-phyboard-izar-rdk-u-boot.dtsi |  4 +++
 arch/arm/dts/k3-am68-r5-phycore-som.dts       |  8 +++++
 board/phytec/phycore_am68x/phycore-am68x.c    | 29 +++++++------------
 configs/phycore_am68x_r5_defconfig            |  1 +
 4 files changed, 23 insertions(+), 19 deletions(-)

diff --git a/arch/arm/dts/k3-am68-phyboard-izar-rdk-u-boot.dtsi b/arch/arm/dts/k3-am68-phyboard-izar-rdk-u-boot.dtsi
index 5ec930f5ade..a6ec4c88cf9 100644
--- a/arch/arm/dts/k3-am68-phyboard-izar-rdk-u-boot.dtsi
+++ b/arch/arm/dts/k3-am68-phyboard-izar-rdk-u-boot.dtsi
@@ -26,6 +26,10 @@
 		spi0 = &ospi0;
 		spi1 = &ospi1;
 	};
+
+	memory@80000000 {
+		bootph-pre-ram;
+	};
 };
 
 &wkup_i2c0 {
diff --git a/arch/arm/dts/k3-am68-r5-phycore-som.dts b/arch/arm/dts/k3-am68-r5-phycore-som.dts
index 0b24da09a77..eaafcfe9d79 100644
--- a/arch/arm/dts/k3-am68-r5-phycore-som.dts
+++ b/arch/arm/dts/k3-am68-r5-phycore-som.dts
@@ -58,6 +58,14 @@
 		bootph-pre-ram;
 	};
 
+	memory@80000000 {
+		device_type = "memory";
+		/* 4 GB RAM */
+		reg = <0x00 0x80000000 0x00 0x80000000>,
+			<0x08 0x80000000 0x00 0x80000000>;
+		bootph-pre-ram;
+	};
+
 	vcc_3v3: regulator-vcc-3v3 {
 		/* Output of SiC431 */
 		compatible = "regulator-fixed";
diff --git a/board/phytec/phycore_am68x/phycore-am68x.c b/board/phytec/phycore_am68x/phycore-am68x.c
index 581ebde8276..f9667618a05 100644
--- a/board/phytec/phycore_am68x/phycore-am68x.c
+++ b/board/phytec/phycore_am68x/phycore-am68x.c
@@ -31,13 +31,11 @@ int board_init(void)
 
 int dram_init(void)
 {
-#ifdef CONFIG_PHYS_64BIT
-	gd->ram_size = 0x100000000;
-#else
-	gd->ram_size = 0x80000000;
-#endif
-
-	return 0;
+	s32 ret;
+	ret = fdtdec_setup_mem_size_base();
+	if (ret)
+		printf("Error setting up mem size and base. %d\n", ret);
+	return ret;
 }
 
 phys_size_t board_get_usable_ram_top(phys_size_t total_size)
@@ -52,20 +50,13 @@ phys_size_t board_get_usable_ram_top(phys_size_t total_size)
 }
 
 int dram_init_banksize(void)
-{
-	/* Bank 0 declares the memory available in the DDR low region */
-	gd->bd->bi_dram[0].start = CFG_SYS_SDRAM_BASE;
-	gd->bd->bi_dram[0].size = 0x7fffffff;
-	gd->ram_size = 0x80000000;
+{	s32 ret;
 
-#ifdef CONFIG_PHYS_64BIT
-	/* Bank 1 declares the memory available in the DDR high region */
-	gd->bd->bi_dram[1].start = CFG_SYS_SDRAM_BASE1;
-	gd->bd->bi_dram[1].size = 0x7fffffff;
-	gd->ram_size = 0x100000000;
-#endif
+	ret = fdtdec_setup_memory_banksize();
+	if (ret)
+		printf("Error setting up memory banksize. %d\n", ret);
 
-	return 0;
+	return ret;
 }
 
 #if defined(CONFIG_SPL_LOAD_FIT)
diff --git a/configs/phycore_am68x_r5_defconfig b/configs/phycore_am68x_r5_defconfig
index c81b21c78a2..13e71853408 100644
--- a/configs/phycore_am68x_r5_defconfig
+++ b/configs/phycore_am68x_r5_defconfig
@@ -181,3 +181,4 @@ CONFIG_FS_FAT_MAX_CLUSTSIZE=16384
 CONFIG_PANIC_HANG=y
 CONFIG_LIB_RATIONAL=y
 CONFIG_SPL_LIB_RATIONAL=y
+CONFIG_NR_DRAM_BANKS=2
-- 
2.25.1

